colossalai/pipeline/policy/bert.py

```diff
@@ -285,53 +305,74 @@ def bert_for_pretraining_forward(
+                                 input_ids,
+                                 attention_mask=attention_mask,
+                                 token_type_ids=token_type_ids,
+                                 position_ids=position_ids,
+                                 head_mask=head_mask,
+                                 inputs_embeds=inputs_embeds,
+                                 output_attentions=output_attentions,
+                                 output_hidden_states=output_hidden_states,
+                                 return_dict=return_dict,
+                                 stage_manager=stage_manager,
+                                 hidden_states=hidden_states if hidden_states is not None else None)
+    past_key_values = None
+    all_hidden_states = None
+    all_self_attentions = None
+    all_cross_attentions = None
+    hidden_states = outputs[0]
+
     if stage_manager.is_last_stage():
+        sequence_output, pooled_output = outputs[:2]
         prediction_scores, seq_relationship_score = self.cls(sequence_output, pooled_output)
+        # the last stage for pretraining model
+        total_loss = None
+        if labels is not None and next_sentence_label is not None:
+            loss_fct = CrossEntropyLoss()
+            masked_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
+            next_sentence_loss = loss_fct(seq_relationship_score.view(-1, 2), next_sentence_label.view(-1))
+            total_loss = masked_lm_loss + next_sentence_loss
 
-    total_loss = None
-    if labels is not None and next_sentence_label is not None:
-        loss_fct = CrossEntropyLoss()
-        masked_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
-        next_sentence_loss = loss_fct(seq_relationship_score.view(-1, 2), next_sentence_label.view(-1))
-        total_loss = masked_lm_loss + next_sentence_loss
-
-    if not return_dict:
-        output = (prediction_scores, seq_relationship_score) + outputs[2:]
-        return ((total_loss,) + output) if total_loss is not None else output
+        if not return_dict:
+            output = (prediction_scores, seq_relationship_score) + outputs[2:]
+            return ((total_loss,) + output) if total_loss is not None else output
+
+        return BertForPreTrainingOutput(
+            loss=total_loss,
+            prediction_logits=prediction_scores,
+            seq_relationship_logits=seq_relationship_score,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
 
-    return BertForPreTrainingOutput(
-        loss=total_loss,
-        prediction_logits=prediction_scores,
-        seq_relationship_logits=seq_relationship_score,
-        hidden_states=outputs.hidden_states,
-        attentions=outputs.attentions,
-    )
+    else:
+        if not return_dict:
+            logger.warning_once('The output of intermediate stage should always be a dict')
```

Comment:
Warning is not needed, since intermdiate outputs are transparent to users.

Reply:
done